[Sema] [AutoDiff] Derive `@_fieldwiseProductSpace` when tangent/cotangent is Self

rxwei opened the pull request:
In `Differentiable` derived conformances, fix a bug where `@_fieldwiseProductSpace` is not added to derived typealiases when `TangentVector` and/or `CotangentVector` are inferred as `Self`. Without this change, field access `x.a` could not be differentiated.

```swift
struct AdditiveTangentIsSelf : AdditiveArithmetic, Differentiable {
  var a: Float
}
let _: @autodiff (AdditiveTangentIsSelf) -> Float = { x in
  x.a + x.a
}
```

rxwei committed:
[Sema] [AutoDiff] Derive `@_fieldwiseProductSpace` when tangent/cotangent is Self.

rxwei commented:
@swift-ci please test tensorflow@swift-ci please test tensorflow

rxwei committed:
Remove mapTypeOutOfContext()

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on test/AutoDiff/derived_differentiable_properties.swift:
```diff
@@ -1,12 +1,27 @@
-// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s
+// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s --check-prefix=CHECK-AST
```
Is the first RUN line meant to perform `-print-ast`?

rxwei replied:
Meant to test the AST printed in SIL.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow macOS

dan-zheng commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -374,6 +374,7 @@ ERROR(autodiff_unsupported_type,none,
       "differentiating '%0' is not supported yet", (Type))
 ERROR(autodiff_function_not_differentiable,none,
       "function is not differentiable", ())
+// TODO: Change this to a note.
```
Do you want to change this to a note?

rxwei replied:
No, here's why. The `DifferentiationInvoker` infra keeps track of how each differentiation task is created. Only the diagnostic on the very bottom of the invocation stack should be an error. 

```swift
func f0(x) {
  f1()
}
func f1(x) {
  f2(x)
}
func f2(x) {
  f3(x) // non-differentiable
}
derivative(of: f0)
```

If expression `f3()` is not differentiable, we should **not** emit an error at `f3()` because it's not the place where the user requested the derivative and the user should not be told to change the code here before knowing what's actually causing differentiation to occur. Instead, we emit notes along the invocation stack until we reach the user's differentiation request, which is always the bottom of the invocation stack.

```swift
func f0(x) {
  f1() // note: when differentiating this function call
}
func f1(x) {
  f2(x) // note: when differentiating this function call
}
func f2(x) {
  f3(x) // note: expression is not differentiable
}
derivative(of: f0) // error: f0 is not differentiable
```

As to why this cannot be changed to a note today: We used to have `DifferentialOperator` as a case in `DifferentiationInvoker`, and the differential operator was, as you know, `#gradient`. When we switched over to generalized differentiability™, there is no more compiler-known differential operator -- function conversion to `@autodiff` is the __only__ formal way to trigger differentiation in expressions. We didn't add a case for function conversion in `DifferentiationInvoker`, so the user request is not recorded, and there is no error at the user request. If we change this error to a note today, the pass pipeline would assume there's no error, keep running other passes on partially/incorrectly transformed code, and cause crashers or undefined runtime behavior.

dan-zheng replied:
Thanks for the detailed info! Definitely interested in improving AD diagnostics sometime.

dan-zheng reviewed (approved):
Good catch!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Use boilerplate code on synthesized typealias declaration.

The crucial line is copying formal access from nominal, which should
fix the failing test.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
